helix-core/src/main/java/org/apache/helix/cloud/topology/FaultZoneBasedVirtualGroupAssignmentAlgorithm.java

```diff
@@ -0,0 +1,195 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+import org.apache.commons.math3.util.Pair;
+
+import static org.apache.helix.util.VirtualTopologyUtil.computeVirtualGroupId;
+
+/**
+ * A virtual group assignment algorithm that assigns zones and their instances to virtual groups
+ * a way that preserves existing zone-to-group assignments whenever possible, and balances any
+ * remaining unassigned zones across the least-loaded groups. If the requested number of groups
+ * differs from the existing assignment, a new distribution is computed. Otherwise, if a zone
+ * already exists in the provided assignment, all its instances (including newly discovered ones)
+ * are placed in the same group, ensuring no zone is split across multiple virtual groups.
+ */
+public class FaultZoneBasedVirtualGroupAssignmentAlgorithm implements VirtualGroupAssignmentAlgorithm {
+
+  private static final FaultZoneBasedVirtualGroupAssignmentAlgorithm _instance =
+      new FaultZoneBasedVirtualGroupAssignmentAlgorithm();
+
+  private FaultZoneBasedVirtualGroupAssignmentAlgorithm() {
+  }
+
+  public static FaultZoneBasedVirtualGroupAssignmentAlgorithm getInstance() {
+    return _instance;
+  }
+
+  @Override
+  public Map<String, Set<String>> computeAssignment(int numGroups, String virtualGroupName,
+      Map<String, Set<String>> zoneMapping, Map<String, Set<String>> virtualZoneMapping) {
+    // Build instance-to-zone mapping for quick zone lookups.
+    Map<String, String> instanceToZoneMapping = new HashMap<>();
+    for (Map.Entry<String, Set<String>> entry : zoneMapping.entrySet()) {
+      for (String instance : entry.getValue()) {
+        instanceToZoneMapping.put(instance, entry.getKey());
+      }
+    }
+    // Collect all instances from zoneMapping
+    Set<String> allInstances = zoneMapping.values()
+        .stream()
+        .flatMap(Set::stream)
+        .collect(Collectors.toSet());
+
+    // 1. If the number of requested virtual groups differs from the current assignment size,
+    //    we must do a fresh assignment (the existing distribution is invalid).
+    if (numGroups != virtualZoneMapping.size()) {
+      Map<String, Set<String>> newAssignment = new HashMap<>();
+      for (int i = 0; i < numGroups; i++) {
+        newAssignment.put(computeVirtualGroupId(i, virtualGroupName), new HashSet<>());
+      }
+
+      // Assign all instances from scratch in a balanced manner.
+      return distributeUnassignedZones(newAssignment, allInstances, instanceToZoneMapping);
+    }
+
+    // Build a deep copy of the current assignment to avoid mutating it directly.
+    Map<String, Set<String>> updatedAssignment = deepCopy(virtualZoneMapping);
```

Comment:
we also build a deep copy in distributeUnassignedZones. I think either this or the one in distributeUnassignedZones can be skipped. (Need to rearrange the code though... for example, moving this in line 75 and remove the one in distributeUnassignedZones)